File: solutions-by-language/java/1792_maximum_average_pass_ratio.java
import java.util.PriorityQueue;

class Solution {
    static class ClassRatio {
        int pass;
        int total;

        ClassRatio(int p , int t) {
            this.pass = p;
            this.total = t;
        }

        void addStudent() {
            pass++;
            total++;
        }

        double ratio () {
            return (double) pass/ total;
        }

        double gain() {
            return ((double) (pass + 1) / (total + 1)) - ratio();
        }
    }
    
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        PriorityQueue<ClassRatio> maxHeap = new PriorityQueue<> ((a,b) -> Double.compare(b.gain(), a.gain()));
        
        for (int [] cls :classes) {
            maxHeap.offer(new ClassRatio(cls[0], cls[1]));
        }

        for(int i = 0 ; i < extraStudents; i++){
            ClassRatio best = maxHeap.poll();
            best.addStudent();
            maxHeap.offer(best);
        }

        double total = 0.0;
        for(ClassRatio cls: maxHeap) {
            total += cls.ratio();
        }
        return total / classes.length;
    }

}

File: solutions-by-language/cpp/1792_maximum_average_pass_ratio.cpp
class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        auto gain = [](int p, int t) {
            return (double)(p + 1) / (t + 1) - (double)p / t;
        };

        priority_queue<pair<double, pair<int, int>>> pq;

        for (auto& cls : classes) {
            pq.push({gain(cls[0], cls[1]), {cls[0], cls[1]}});
        }

        while (extraStudents--) {
            auto [g, pt] = pq.top(); pq.pop();
            int p = pt.first + 1;
            int t = pt.second + 1;
            pq.push({gain(p, t), {p, t}});
        }

        double total = 0.0;
        while (!pq.empty()) {
            auto [_, pt] = pq.top(); pq.pop();
            total += (double)pt.first / pt.second;
        }

        return total / classes.size();
    }
};